internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,500 @@
+	"k8s.io/apiserver/pkg/endpoints/request"
+	rbacinternal "k8s.io/kubernetes/pkg/apis/rbac"
+	rbacv1helpers "k8s.io/kubernetes/pkg/apis/rbac/v1"
+	rbacregistry "k8s.io/kubernetes/pkg/registry/rbac"
+	"k8s.io/kubernetes/pkg/registry/rbac/validation"
+	rbac "k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type PreAuthorizer interface {
+	PreAuthorize(ctx context.Context, manifestManager user.Info, manifestReader io.Reader) ([]ScopedPolicyRules, error)
+}
+
+type ScopedPolicyRules struct {
+	Namespace    string
+	MissingRules []rbacv1.PolicyRule
+}
+
+var (
+	collectionVerbs = []string{"list", "watch", "create"}
+	objectVerbs     = []string{"get", "patch", "update", "delete"}
+)
+
+type rbacPreAuthorizer struct {
+	authorizer   authorizer.Authorizer
+	ruleResolver validation.AuthorizationRuleResolver
+	restMapper   meta.RESTMapper
+}
+
+func NewRBACPreAuthorizer(cl client.Client) PreAuthorizer {
+	return &rbacPreAuthorizer{
+		authorizer:   newRBACAuthorizer(cl),
+		ruleResolver: newRBACRulesResolver(cl),
+		restMapper:   cl.RESTMapper(),
+	}
+}
+
+// PreAuthorize validates whether the current user/request satisfies the necessary permissions
+// as defined by the RBAC policy. It examines the user’s roles, resource identifiers, and
+// the intended action to determine if the operation is allowed.
+//
+// Return Value:
+//   - nil: indicates that the authorization check passed and the operation is permitted.
+//   - non-nil error: indicates that an error occurred during the permission evaluation process
+//     (for example, a failure decoding the manifest or other internal issues). If the evaluation
+//     completes successfully but identifies missing rules, then a nil error is returned along with
+//     the list (or slice) of missing rules. Note that in some cases the error may encapsulate multiple
+//     evaluation failures
+func (a *rbacPreAuthorizer) PreAuthorize(ctx context.Context, manifestManager user.Info, manifestReader io.Reader) ([]ScopedPolicyRules, error) {
+	var allMissingPolicyRules = []ScopedPolicyRules{}
+	dm, err := a.decodeManifest(manifestReader)
+	if err != nil {
+		return nil, err
+	}
+	attributesRecords := dm.asAuthorizationAttributesRecordsForUser(manifestManager)
+
+	var preAuthEvaluationErrors []error
+	missingRules, err := a.authorizeAttributesRecords(ctx, attributesRecords)
+	if err != nil {
+		preAuthEvaluationErrors = append(preAuthEvaluationErrors, err)
+	}
+
+	ec := escalationChecker{
+		authorizer:        a.authorizer,
+		ruleResolver:      a.ruleResolver,
+		extraClusterRoles: dm.clusterRoles,
+		extraRoles:        dm.roles,
+	}
+
+	for _, obj := range dm.rbacObjects() {
+		if err := ec.checkEscalation(ctx, manifestManager, obj); err != nil {
+			// In Kubernetes 1.32.2 the specialized PrivilegeEscalationError is gone.
+			// Instead, we simply collect the error.
+			preAuthEvaluationErrors = append(preAuthEvaluationErrors, err)
+		}
```

Comment:
We've got a problem here. `ec.checkEscalation` also returns missing rules, but:
1. They are embedded in the returned `err`
2. The returned error is a simple string combines information about missing rules and separate evaluation errors.

We need the error returned by `ec.checkEscalation` to be something we can type assert on and extract out the missing rules, so that we can add to our missing rule set and still collect the separate evaluation errors that are possible. That was the purpose of `PrivilegeEscalationError` in my PoC.

Reply:
Took a stab at handling this, @bentito please review

Reply:
Yeah I think it's a decent change, better UX in collecting the escalation errors. It's still doing string parsing here: 

https://github.com/operator-framework/operator-controller/pull/1858/files#diff-e6900bff718bc092b24c88419aec29e4ac605e87a733a6484aafbf42e5eb89dcR543

and it seemed this is where @joelanford was looking for a new, type-assertable error structure?

Reply:
marking this resolved as it's tracked in another comment thread